### PR TITLE
Service Route Peerings update based on Service Policy module tests

### PR DESCRIPTION
**Pull request by mmudigon**
A] Service Route Peering updates include:

1. Check for authentication errors from dcnm_send to reset connection and retry
2. Note down critical responses in "result"
3. Override can now be performed without any config or with a new peering
4. Delete can now be performed with no config, with node name and node and peering name
5. Black and Flake8 changes

B] Service Policy module first commit that includes
     - backend module
     - IT and UT code

### CODE REVIEW

**Comment by mikewiebe** [comment]
@mmudigon Please update the changelog for these new feature modules.

File: plugins/modules/dcnm_service_policy.py [hunk 1]
```diff
@@ -0,0 +1,1962 @@
...
+    config:
+      - node_name: it-sn-1
+        rp_name: it-fw-rp1
+
+      - node_name: it-sn-2
+        rp_name: it-fw-rp2
+
+4. Without config
+
+   Deletes all service policies on the given fabric and attached fabirc
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:305]: `/fabirc/fabric/`

File: plugins/modules/dcnm_service_policy.py [hunk 2]
```diff
@@ -0,0 +1,1962 @@
...
+
+- name: Delete service policies without config
+  cisco.dcnm.dcnm_service_policy:
+    fabric: test_fabirc
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:309]: `/fabirc/fabric/`. Check for other places where this typo might exist.

File: plugins/modules/dcnm_service_policy.py [hunk 3]
```diff
@@ -0,0 +1,1962 @@
...
+    attached_fabric: external
+    state: deleted
+
+OVERRIDE SERVICE POLICIES
+=========================
+
+- name: Override and delete all service policies
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:316]: Just for clarification, we are supporting the fact that `state: overridden` without the `config:` keyword behaves just like `state: deleted` without the `config:` keyword right?  

Not sure I would actually document this case.  While it makes sense it's not something I would recommend customers use.  We should just document the `state: deleted` case.

File: plugins/modules/dcnm_service_policy.py [hunk 4]
```diff
@@ -0,0 +1,1962 @@
...
+  cisco.dcnm.dcnm_service_policy:
+    fabric: test_fabric
+    attached_fabric: external
+    state: overridden
+
+- name: Override all existing service policies with a new one
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:322]: For the sake of clarify please add a note in this example that service policy `service_policy_1` will be created or replaced and all other service policies in `test_fabric` and `external` will be deleted.

File: plugins/modules/dcnm_service_policy.py [hunk 5]
```diff
@@ -0,0 +1,1962 @@
...
+          next_hop_option:  none
+          acl_name: fwd_acl_555
+          rev_acl_name: rev_acl_555
+          route_map_num: 555
+          rev_route_map_num: 556
+
+REPLACE SERVICE POLICIES
+========================
+
+- name: Replace some of the objects in already created service policy
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:353]: The way I read this it seems like replaced is more like a merge.  I would state that `service_policy_1` will be replaced on DCNM with the ansible playbook version.

File: plugins/modules/dcnm_service_policy.py [hunk 6]
```diff
@@ -0,0 +1,1962 @@
...
+from ansible_collections.cisco.dcnm.plugins.module_utils.network.dcnm.dcnm import (
+    dcnm_send,
+    validate_list_of_dicts,
+    dcnm_reset_connection,
+)
+
+from datetime import datetime
+
+
+# Service Policy Class object which includes all the required methods and data to configure and maintain Roue peering objects
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:413]: `/Roue/Route/`

But should it read Service Policy Objects instead?

File: plugins/modules/dcnm_service_policy.py [hunk 7]
```diff
@@ -0,0 +1,1962 @@
...
+        sp_spec = dict(
+            name=dict(required=True, type="str"),
+            node_name=dict(required=True, type="str"),
+            rp_name=dict(required=True, type="str"),
+            src_vrf=dict(required=True, type="str"),
+            dest_vrf=dict(required=True, type="str"),
+            src_network=dict(required=True, type="str"),
+            dest_network=dict(required=True, type="str"),
+            next_hop=dict(required=True, type="ipv4"),
+            rev_next_hop=dict(required=True, type="ipv4"),
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:572]: Looks like `rev_next_hop` is not listed as required in the documentation.  Please double check all required parameters in the code to make sure the docs match.

File: plugins/modules/dcnm_service_policy.py [hunk 8]
```diff
@@ -0,0 +1,1962 @@
...
+            resp = dcnm_send(self.module, "GET", path)
+
+            if resp and (resp["RETURN_CODE"] == 200) and resp["DATA"]:
+                return resp["DATA"].get("type", None)
+            else:
+                self.dcnm_sp_check_unauthorized_error_in_resp(resp)
+                time.sleep(10)
+                continue
+        self.changed_dict[0]["debugs"].append({"GET_SNODE_TYPE": resp})
+        return "Firewall"
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:694]: This return confuses me.  Why are we returning a hardcoded value here?

File: plugins/modules/dcnm_service_policy.py [hunk 9]
```diff
@@ -0,0 +1,1962 @@
...
+        sp_payload["policyName"] = sp["name"]
+        sp_payload["peeringName"] = sp["rp_name"]
+        sp_payload["sourceVrfName"] = sp["src_vrf"]
+        sp_payload["destinationVrfName"] = sp["dest_vrf"]
+        sp_payload["sourceNetwork"] = sp["src_network"]
+        sp_payload["destinationNetwork"] = sp["dest_network"]
+        sp_payload["reverseEnabled"] = sp["reverse"]
+        sp_payload["policyTemplateName"] = "service_pbr"
+        sp_payload["nextHopIp"] = sp["next_hop"]
+        if sp["reverse"]:
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:729]: Is `sp` a dict?  If so should you use `sp.get("reverse")` just to be safe?

File: plugins/modules/dcnm_service_policy.py [hunk 10]
```diff
@@ -0,0 +1,1962 @@
...
+                if (
+                    have["nvPairs"]["REVERSE"] == "True"
+                    or have["nvPairs"]["REVERSE"] == "true"
+                ):
+                    have["nvPairs"]["REVERSE"] = True
+                if (
+                    have["nvPairs"]["REVERSE"] == "False"
+                    or have["nvPairs"]["REVERSE"] == "false"
+                ):
+                    have["nvPairs"]["REVERSE"] = False
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_policy.py:1189]: Suggest the following simplification for the logic above:

```python
import re
if key == "REVERSE":
    if re.search(r'true', have["nvPairs"]["REVERSE"].lower())
        have["nvPairs"]["REVERSE"] = True
    if re.search(r'false', have["nvPairs"]["REVERSE"].lower())
        have["nvPairs"]["REVERSE"] = False
```

**Review by mikewiebe — CHANGES_REQUESTED** [review]
Lot of work here @mmudigon ! :). Still reviewing but here is my first set of comments.

File: plugins/httpapi/dcnm.py [hunk 11]
```diff
@@ -61,24 +61,15 @@ def login(self, username, password):
...
 
-        try:
-            response, response_data = self.connection.send(path, data, method=method, headers=self.headers, force_basic_auth=True)
-            response_value = self._get_response_value(response_data)
-            self.connection._auth = {'Dcnm-Token': self._response_to_json(response_value)['Dcnm-Token']}
-
-        except Exception as e:
-            msg = 'Error on attempt to connect and authenticate with DCNM controller: {}'.format(e)
-            raise ConnectionError(self._return_info(None, method, path, msg))
+        response, response_data = self.connection.send(path, data, method=method, headers=self.headers, force_basic_auth=True)
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/httpapi/dcnm.py:64]: Agree. This was kept this way to make sure we can catch the other error that was seen while testing. I can revert the changes for now.

File: plugins/modules/dcnm_service_policy.py [hunk 12]
```diff
@@ -0,0 +1,1962 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__author__ = "Mallik Mudigonda"
+
+DOCUMENTATION = """
+---
+module: dcnm_service_policy
+short_description: dcnm ansible module for managing service policies.
+version_added: "1.1.0"
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:23]: Sure, I will update all 3 modules to 1.2.0

File: plugins/modules/dcnm_service_policy.py [hunk 13]
```diff
@@ -0,0 +1,1962 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__author__ = "Mallik Mudigonda"
+
+DOCUMENTATION = """
+---
+module: dcnm_service_policy
+short_description: dcnm ansible module for managing service policies.
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:22]: Done

File: plugins/modules/dcnm_service_policy.py [hunk 14]
```diff
@@ -0,0 +1,1962 @@
...
+version_added: "1.1.0"
+description:
+    - dcnm ansible module for creating, deleting, querying and modifying service policies
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:25]: Done

File: plugins/modules/dcnm_service_policy.py [hunk 15]
```diff
@@ -0,0 +1,1962 @@
...
+    elements: dict
+    suboptions:
+      name:
+        description:
+          - a unique name which identifies the service policy
+        type: str
+        required: true
+      src_vrf:
+        description:
+          - name of the source vrf for which the service policy is applicable
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:75]: yes, agree

File: plugins/modules/dcnm_service_policy.py [hunk 16]
```diff
@@ -0,0 +1,1962 @@
...
+        description:
+          - name of the destination network for which the service policy is applicable
+        type: str
+        required: true
+      next_hop:
+        description:
+          - next hop ip address to be used in source to network direction
+          - NOTE: This must exactly match the next hop IP configured for the route
+                  peering associated with this policy
+        type: ipv4
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:98]: No. We have only IPV4 support

File: plugins/modules/dcnm_service_policy.py [hunk 17]
```diff
@@ -0,0 +1,1962 @@
...
+        required: false
+        default: ''
+      rev_next_hop:
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:101]: Sure. Will change that. I have also sent a mail to Shangxin regarding the requirement of these fields in playbook.He said he will get back. We may remove these if not required.

File: plugins/modules/dcnm_service_policy.py [hunk 18]
```diff
@@ -0,0 +1,1962 @@
...
+            required: false
+            choices:  [none, drop-on-fail, drop]
+            default: none
+          acl_name:
+            description:
+              - Name of the ACL in the forward direction
+            type: str
+            required: false
+            default: will be auto-generated by DCNM
+          rev_acl_name:
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:156]: Same as for rev_next_hop

File: plugins/modules/dcnm_service_policy.py [hunk 19]
```diff
@@ -0,0 +1,1962 @@
...
+            required: false
+            default: will be auto-generated by DCNM
+          route_map_num:
+            description:
+              - route map match number
+            type: int
+            required: false
+            default: will be auto-generated by DCNM
+            choices: [Min:1, Max:65535)]
+          rev_route_map_num:
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:169]: Same as other two

File: plugins/modules/dcnm_service_policy.py [hunk 20]
```diff
@@ -0,0 +1,1962 @@
...
+            description:
+              - route map match number for reverse direction
+            type: int
+            required: false
+            default: will be auto-generated by DCNM
+            choices: [Min:1, Max:65535)]
+"""
+
+EXAMPLES = """
+
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:179]: Sure

File: plugins/modules/dcnm_service_policy.py [hunk 21]
```diff
@@ -0,0 +1,1962 @@
...
+States:
+This module supports the following states:
+
+Merged:
+  Service Peerings defined in the playbook will be merged into the target fabric.
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:184]: No, it should be Service Policy. Will replace all.

File: plugins/modules/dcnm_service_policy.py (hunk 1 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:305]: Done

File: plugins/modules/dcnm_service_policy.py (hunk 2 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:309]: Done

File: plugins/modules/dcnm_service_policy.py (hunk 3 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:316]: Ok. Will remove this from EXAMPLES. I will only include the 'overridden' with a policy alone

File: plugins/modules/dcnm_service_policy.py (hunk 4 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:322]: Done

File: plugins/modules/dcnm_service_policy.py (hunk 5 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:353]: Done

File: plugins/modules/dcnm_service_policy.py (hunk 6 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:413]: I have replaced Route Peerings with Service Policies in a few places. As part of that this too would have been replaced.

File: plugins/modules/dcnm_service_policy.py (hunk 7 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:572]: Done

File: plugins/modules/dcnm_service_policy.py (hunk 8 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:694]: Should not. It should be from the response. Removed it.

File: plugins/modules/dcnm_service_policy.py (hunk 9 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:729]: We can. But sp["reverse"] is defaulted to True in Validate function if it is not given in the playbook. So we are safe.

File: plugins/modules/dcnm_service_policy.py (hunk 10 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:1189]: or how about have["nvPairs"]["REVERSE].lower() == "true"

File: README.md [hunk 22]
```diff
@@ -36,6 +36,8 @@ Name | Description
 [cisco.dcnm.dcnm_policy](https://github.com/CiscoDevNet/ansible-dcnm/blob/main/docs/cisco.dcnm.dcnm_policy_module.rst)|DCNM Ansible Module for managing policies.
 [cisco.dcnm.dcnm_rest](https://github.com/CiscoDevNet/ansible-dcnm/blob/main/docs/cisco.dcnm.dcnm_rest_module.rst)|Send REST API requests to DCNM controller.
 [cisco.dcnm.dcnm_service_node](https://github.com/CiscoDevNet/ansible-dcnm/blob/main/docs/cisco.dcnm.dcnm_service_node_module.rst)|Create/Modify/Delete service node based on type and attached interfaces from a DCNM managed VXLAN fabric.
+[cisco.dcnm.dcnm_service_policy](https://github.com/CiscoDevNet/ansible-dcnm/blob/main/docs/cisco.dcnm.dcnm_service_policy_module.rst)|dcnm ansible module for managing service policies.
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply README.md:39]: Ok

File: meta/runtime.yml [hunk 23]
```diff
@@ -3,7 +3,7 @@
 # will redirect to a common action plugin called cisco.dcnm.dcnm
 #
 #---
-#requires_ansible: '>=2.9.10,<2.11'
+requires_ansible: '>=2.9.10,<2.11'
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply meta/runtime.yml:6]: Not sure. But when we ran collection_prep we saw some error. That's when we corrected this.

File: plugins/modules/dcnm_service_route_peering.py [hunk 24]
```diff
@@ -126,7 +126,8 @@
             description:
               - Vlan Id for the inside network
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:127]: Done

File: plugins/modules/dcnm_service_policy.py [hunk 25]
```diff
@@ -0,0 +1,1962 @@
...
+        required: true
+      dest_network:
+        description:
+          - name of the destination network for which the service policy is applicable
+        type: str
+        required: true
+      next_hop:
+        description:
+          - next hop ip address to be used in source to network direction
+          - NOTE: This must exactly match the next hop IP configured for the route
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:96]: The error is generated. It specifically says that the address does not match what is configured in route peering

File: plugins/modules/dcnm_service_policy.py [hunk 26]
```diff
@@ -0,0 +1,1962 @@
...
+                  peering associated with this policy
+        type: ipv4
+        required: false
+        default: ''
+      rev_next_hop:
+        description:
+          - reverse next hop ip address to be used in network to source direction
+          - NOTE: This must exactly match the reverse next hop IP configured for the route
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_policy.py:104]: Same as previous

File: plugins/modules/dcnm_service_policy.py (hunk 10 above)
Comment (mikewiebe) [reply plugins/modules/dcnm_service_policy.py:1189]: Sure, that also works.

File: plugins/modules/dcnm_service_route_peering.py [hunk 27]
```diff
@@ -3506,7 +4321,9 @@ def main():
         debug=dict(required=False, type="bool", default=False),
     )
 
-    module = AnsibleModule(argument_spec=element_spec, supports_check_mode=True)
+    module = AnsibleModule(
+        argument_spec=element_spec, supports_check_mode=True
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_route_peering.py:4325]: I think we discussed this a few weeks back but looks like `check_mode` is still specified as a module argument.  This is an ansible infra level arg so we need to update the this module and `service_node` and `service_policy` modules if they have the same issue.

@praveenramoorthy recently made a similar update/fix for dcnm_vrf and dcnm_network modules

See: https://github.com/CiscoDevNet/ansible-dcnm/pull/88/files

**Comment by mikewiebe** [comment]
Looks like there is a merge conflict now.  Please resolve.

**Review by mikewiebe — APPROVED** [review]

**Pull request merged**